File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(candidates);
        fun(0, target, candidates, new ArrayList<>(), ans);
        return ans;
    }

    private void fun(int index, int target, int[] arr, List<Integer> subset, List<List<Integer>> ans){
        if(target == 0){
            ans.add(new ArrayList<>(subset));
            return;
        }
        for(int i=index; i<arr.length;i++){
            if(i > index && arr[i] == arr[i-1])
                continue;
            if(arr[i] > target)
                break;

            subset.add(arr[i]);
            fun(i + 1, target - arr[i], arr, subset, ans);
            subset.remove(subset.size() - 1);
        }
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:

    def fun(self, index, target, arr, subset, ans):
        if target == 0:
            print(subset)
            ans.append(subset[:])
            return

        for i in range(index, len(arr)): 
            if i > index and arr[i] == arr[i-1]:
                continue
            if arr[i] > target:
                break

            subset.append(arr[i])
            self.fun(i + 1, target - arr[i], arr, subset, ans)
            subset.pop()
        
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        candidates.sort()
        self.fun(0, target, candidates, [], ans)
        return ans